FIAT/discontinuous_lagrange.py

```diff
@@ -5,7 +5,110 @@
+    #      [1, 0, 2],
+    #      [1, 1, 1],    (0,1,2)(1,1,1)(2,1,0)
+    #      [1, 2, 0],
+    #      [2, 0, 1],    (0,0,3)(1,0,2)(2,0,1)(3,0,0)
+    #      [2, 1, 0],
+    #      [3, 0, 0]]    Lattice points that a represents
+    # a.shape[0] = number of DoFs
+    # a.shape[1] = dim + 1
+    # sum of each row = degree (bary centric lattice coordinates)
+    index_perms = make_index_permutations(dim + 1)
+    # index_perms = [[0, 1, 2],
+    #                [0, 2, 1],
+    #                [1, 0, 2],
+    #                [1, 2, 0],
+    #                [2, 0, 1],
+    #                [2, 1, 0]]
+    g0 = dim - (a == 0).astype(int).sum(axis=1)
+    # g0 = [ 0, 1, 1, 0, 1, 2, 1, 1, 1, 0]  # 0 for vertices
+    #                                       # 1 for edges
+    #                                       # 2 for cell
+    g1 = np.zeros_like(g0)
+    for d in range(dim + 1):
+        on_facet_d = (a[:, d] == 0).astype(int)
+        g1 += d * on_facet_d
+    # The above logic is consistent with the FIAT entity numbering
+    # convention ("entities that are not incident to vertex 0 are
+    # numbered first, then ..."), but vertices are not numbered using
+    # the same logic in FIAT, so we need to fix:
+    g1[g0 == 0] = -g1[g0 == 0]
+    # g1 = [-3, 2, 2,-2, 1, 0, 0, 1, 0,-1]
+    # For each entity dimension, DoFs that have smaller numbers in g1
+    # will be assigned smaller DG node numbers.
+    # Order first by g0, then by g1, and finally by a (lexicographically as in CG)
+    g0 = g0.reshape((a.shape[0], 1))
+    g1 = g1.reshape((a.shape[0], 1))
+    dg_cg_map = np.lexsort(np.transpose(np.concatenate((a, g1, g0), axis=1)))
+    # dg_cg_map = [ 0, 3, 9, 6, 8, 4, 7, 1, 2, 5]
+    cg_dg_map = np.empty_like(dg_cg_map)
+    for i, im in enumerate(dg_cg_map):
+        cg_dg_map[im] = i
+    # cg_dg_map = [ 0, 7, 8, 1, 5, 9, 3, 6, 4, 2]
+    perms = {}
+    for o, index_perm in enumerate(index_perms):
+        # First compute permutation in CG node numbers in CG order (as we do for CG)
+        perm = np.lexsort(np.transpose(a[:, index_perm]))
+        # Then convert to DG node numbers in DG order:
+        # cg_dg_map[perm]            -> convert CG node numbers to DG node numbers
+        # cg_dg_map[perm][dg_cg_map] -> reorder for DG ordering
+        #
+        # Example:
+        # Under one CCW rotation, a DG element on a physical cell
+        # is mapped to the FIAT reference as:
+        #
+        # 2
+        # 4 6
+        # 3 9 5
+        # 1 8 7 0
+        #
+        # Under the same transformation, a CG element would
+        # be mapped as:
+        #
+        # 9
+        # 8 7
+        # 6 5 4
+        # 3 2 1 0
+        #
+        # In this case we will have:
+        #
+        # perm                       = [3, 6, 8, 9, 2, 5, 7, 1, 4, 0]
+        # cg_dg_map[perm]            = [1, 3, 4, 2, 8, 9, 6, 7, 5, 0]
+        # cg_dg_map[perm][dg_cg_map] = [1, 2, 0, 6, 5, 8, 7, 3, 4, 9]
+        #
```

Comment:
TODO: add comment noting that the sane thing to do is to just number DG dofs on a lattice.

Reply:
Done.